File: java/src/main/java/fr/sdecout/handson/rest/author/AuthorId.java
package fr.sdecout.handson.rest.author;

import jakarta.annotation.Nonnull;

import java.util.UUID;

import static java.util.Objects.requireNonNull;

public record AuthorId(@Nonnull String value) {
    public AuthorId {
        requireNonNull(value);
        if (value.isBlank()) {
            throw new IllegalArgumentException("Author ID must not be blank");
        }
        if (value.length() > 36) {
            throw new IllegalArgumentException("Author ID must not have more than 36 characters");
        }
    }

    public static AuthorId next() {
        return new AuthorId(UUID.randomUUID().toString());
    }

    @Override
    public String toString() {
        return value;
    }
}

File: kotlin/src/main/kotlin/fr/sdecout/handson/rest/author/AuthorId.kt
package fr.sdecout.handson.rest.author

import java.util.*

@JvmInline
value class AuthorId(val value: String) {

    init {
        require(value.isNotBlank()) { "Author ID must not be blank" }
        require(value.length < 37) { "Author ID must not have more than 36 characters" }
    }

    override fun toString() = value

    companion object {
        fun next(): AuthorId = AuthorId(UUID.randomUUID().toString())
    }

}
